File: src/api/client/javascript/usdc-arbitrage-client.js
/**
 * JavaScript client SDK for USDC Arbitrage API.
 */

class USDCArbitrageClient {
  /**
   * Initialize client.
   * @param {string} baseUrl - Base URL for API
   * @param {string} apiVersion - API version
   * @param {string} token - Authentication token
   */
  constructor(baseUrl = 'http://localhost:8000', apiVersion = '1.0', token = null) {
    this.baseUrl = baseUrl;
    this.apiVersion = apiVersion;
    this.token = token;
  }

  /**
   * Get request headers.
   * @returns {Object} - Headers object
   */
  _getHeaders() {
    const headers = {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      'Accept-Version': this.apiVersion,
    };

    if (this.token) {
      headers['Authorization'] = `Bearer ${this.token}`;
    }

    return headers;
  }

  /**
   * Handle API response.
   * @param {Response} response - Fetch response
   * @returns {Promise<Object>} - Response data
   */
  async _handleResponse(response) {
    if (!response.ok) {
      let errorDetail = {};
      try {
        errorDetail = await response.json();
      } catch (e) {
        errorDetail = { detail: await response.text() };
      }
      throw new Error(`API error: ${response.status} - ${JSON.stringify(errorDetail)}`);
    }

    return response.json();
  }

  /**
   * Login to API and get access token.
   * @param {string} username - Username
   * @param {string} password - Password
   * @returns {Promise<Object>} - Login response
   */
  async login(username, password) {
    const url = `${this.baseUrl}/auth/token`;
    const data = {
      username,
      password,
    };

    const response = await fetch(url, {
      method: 'POST',
      headers: this._getHeaders(),
      body: JSON.stringify(data),
    });

    const result = await this._handleResponse(response);

    if (result.access_token) {
      this.token = result.access_token;
    }

    return result;
  }

  /**
   * Refresh access token.
   * @param {string} refreshToken - Refresh token
   * @returns {Promise<Object>} - Refresh response
   */
  async refreshToken(refreshToken) {
    const url = `${this.baseUrl}/auth/refresh`;
    const data = {
      refresh_token: refreshToken,
    };

    const response = await fetch(url, {
      method: 'POST',
      headers: this._getHeaders(),
      body: JSON.stringify(data),
    });

    const result = await this._handleResponse(response);

    if (result.access_token) {
      this.token = result.access_token;
    }

    return result;
  }

  /**
   * List strategies.
   * @param {Object} options - Options
   * @param {boolean} options.activeOnly - Only include active strategies
   * @param {string} options.tag - Filter by tag
   * @param {string} options.search - Search in name and description
   * @param {number} options.skip - Number of records to skip
   * @param {number} options.limit - Maximum number of records
   * @returns {Promise<Array>} - List of strategies
   */
  async listStrategies({
    activeOnly = false,
    tag = null,
    search = null,
    skip = 0,
    limit = 100,
  } = {}) {
    const url = new URL(`${this.baseUrl}/strategies/`);
    url.searchParams.append('active_only', activeOnly);
    url.searchParams.append('skip', skip);
    url.searchParams.append('limit', limit);

    if (tag) {
      url.searchParams.append('tag', tag);
    }

    if (search) {
      url.searchParams.append('search', search);
    }

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    return this._handleResponse(response);
  }

  /**
   * Get strategy by ID.
   * @param {number} strategyId - Strategy ID
   * @param {boolean} includeCode - Include strategy code
   * @returns {Promise<Object>} - Strategy details
   */
  async getStrategy(strategyId, includeCode = false) {
    const url = new URL(`${this.baseUrl}/strategies/${strategyId}`);
    url.searchParams.append('include_code', includeCode);

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    return this._handleResponse(response);
  }

  /**
   * Create a new strategy.
   * @param {Object} request - Strategy creation request
   * @param {string} author - Author name
   * @returns {Promise<Object>} - Created strategy
   */
  async createStrategy(request, author = 'api') {
    const url = new URL(`${this.baseUrl}/strategies/`);
    url.searchParams.append('author', author);

    const response = await fetch(url, {
      method: 'POST',
      headers: this._getHeaders(),
      body: JSON.stringify(request),
    });

    return this._handleResponse(response);
  }

  /**
   * Update an existing strategy.
   * @param {number} strategyId - Strategy ID
   * @param {Object} request - Strategy update request
   * @param {string} author - Author name
   * @returns {Promise<Object>} - Updated strategy
   */
  async updateStrategy(strategyId, request, author = 'api') {
    const url = new URL(`${this.baseUrl}/strategies/${strategyId}`);
    url.searchParams.append('author', author);

    const response = await fetch(url, {
      method: 'PUT',
      headers: this._getHeaders(),
      body: JSON.stringify(request),
    });

    return this._handleResponse(response);
  }

  /**
   * Delete a strategy.
   * @param {number} strategyId - Strategy ID
   * @returns {Promise<Object>} - Deletion response
   */
  async deleteStrategy(strategyId) {
    const url = `${this.baseUrl}/strategies/${strategyId}`;

    const response = await fetch(url, {
      method: 'DELETE',
      headers: this._getHeaders(),
    });

    return this._handleResponse(response);
  }

  /**
   * Run a backtest.
   * @param {Object} request - Backtest request
   * @returns {Promise<Object>} - Backtest response
   */
  async runBacktest(request) {
    const url = `${this.baseUrl}/backtest/`;

    const response = await fetch(url, {
      method: 'POST',
      headers: this._getHeaders(),
      body: JSON.stringify(request),
    });

    return this._handleResponse(response);
  }

  /**
   * Get backtest by ID.
   * @param {number} backtestId - Backtest ID
   * @returns {Promise<Object>} - Backtest details
   */
  async getBacktest(backtestId) {
    const url = `${this.baseUrl}/backtest/${backtestId}`;

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    return this._handleResponse(response);
  }

  /**
   * List backtests.
   * @param {Object} options - Options
   * @param {number} options.strategyId - Filter by strategy ID
   * @param {string} options.status - Filter by status
   * @param {number} options.limit - Maximum number of records
   * @param {number} options.offset - Number of records to skip
   * @returns {Promise<Array>} - List of backtests
   */
  async listBacktests({
    strategyId = null,
    status = null,
    limit = 10,
    offset = 0,
  } = {}) {
    const url = new URL(`${this.baseUrl}/backtest/`);
    url.searchParams.append('limit', limit);
    url.searchParams.append('offset', offset);

    if (strategyId !== null) {
      url.searchParams.append('strategy_id', strategyId);
    }

    if (status !== null) {
      url.searchParams.append('status', status);
    }

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    return this._handleResponse(response);
  }

  /**
   * List backtest results.
   * @param {Object} options - Options
   * @param {number} options.strategyId - Filter by strategy ID
   * @param {string} options.status - Filter by status
   * @param {number} options.skip - Number of records to skip
   * @param {number} options.limit - Maximum number of records
   * @param {string} options.sortBy - Field to sort by
   * @param {string} options.sortOrder - Sort order (asc or desc)
   * @returns {Promise<Array>} - List of backtest results
   */
  async listResults({
    strategyId = null,
    status = null,
    skip = 0,
    limit = 100,
    sortBy = 'created_at',
    sortOrder = 'desc',
  } = {}) {
    const url = new URL(`${this.baseUrl}/results/`);
    url.searchParams.append('skip', skip);
    url.searchParams.append('limit', limit);
    url.searchParams.append('sort_by', sortBy);
    url.searchParams.append('sort_order', sortOrder);

    if (strategyId !== null) {
      url.searchParams.append('strategy_id', strategyId);
    }

    if (status !== null) {
      url.searchParams.append('status', status);
    }

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    return this._handleResponse(response);
  }

  /**
   * Get backtest result by ID.
   * @param {number} resultId - Result ID
   * @returns {Promise<Object>} - Result details
   */
  async getResult(resultId) {
    const url = `${this.baseUrl}/results/${resultId}`;

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    return this._handleResponse(response);
  }

  /**
   * Get transactions for a backtest result.
   * @param {number} resultId - Result ID
   * @param {number} skip - Number of records to skip
   * @param {number} limit - Maximum number of records
   * @returns {Promise<Array>} - List of transactions
   */
  async getResultTransactions(resultId, skip = 0, limit = 100) {
    const url = new URL(`${this.baseUrl}/results/${resultId}/transactions`);
    url.searchParams.append('skip', skip);
    url.searchParams.append('limit', limit);

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    return this._handleResponse(response);
  }

  /**
   * Get position snapshots for a backtest result.
   * @param {number} resultId - Result ID
   * @param {number} skip - Number of records to skip
   * @param {number} limit - Maximum number of records
   * @returns {Promise<Array>} - List of position snapshots
   */
  async getResultPositions(resultId, skip = 0, limit = 100) {
    const url = new URL(`${this.baseUrl}/results/${resultId}/positions`);
    url.searchParams.append('skip', skip);
    url.searchParams.append('limit', limit);

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    return this._handleResponse(response);
  }

  /**
   * Get equity curve for a backtest result.
   * @param {number} resultId - Result ID
   * @returns {Promise<Array>} - Equity curve data
   */
  async getResultEquityCurve(resultId) {
    const url = `${this.baseUrl}/results/${resultId}/equity_curve`;

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    return this._handleResponse(response);
  }

  /**
   * Export data in the requested format.
   * @param {Object} options - Export options
   * @param {string} options.dataType - Data type
   * @param {string} options.format - Export format
   * @param {Object} options.filters - Filters
   * @param {Date} options.startDate - Start date
   * @param {Date} options.endDate - End date
   * @param {Array<string>} options.exchanges - Exchanges
   * @param {Array<string>} options.symbols - Symbols
   * @param {number} options.strategyId - Strategy ID
   * @param {number} options.backtestId - Backtest ID
   * @param {number} options.limit - Maximum number of records
   * @returns {Promise<Blob>} - Exported data
   */
  async exportData({
    dataType,
    format = 'json',
    filters = null,
    startDate = null,
    endDate = null,
    exchanges = null,
    symbols = null,
    strategyId = null,
    backtestId = null,
    limit = 10000,
  }) {
    const url = `${this.baseUrl}/export/data`;
    const data = {
      data_type: dataType,
      format,
    };

    if (filters) {
      data.filters = filters;
    }

    if (startDate) {
      data.start_date = startDate.toISOString();
    }

    if (endDate) {
      data.end_date = endDate.toISOString();
    }

    if (exchanges) {
      data.exchanges = exchanges;
    }

    if (symbols) {
      data.symbols = symbols;
    }

    if (strategyId) {
      data.strategy_id = strategyId;
    }

    if (backtestId) {
      data.backtest_id = backtestId;
    }

    if (limit) {
      data.limit = limit;
    }

    const response = await fetch(url, {
      method: 'POST',
      headers: this._getHeaders(),
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      let errorDetail = {};
      try {
        errorDetail = await response.json();
      } catch (e) {
        errorDetail = { detail: await response.text() };
      }
      throw new Error(`API error: ${response.status} - ${JSON.stringify(errorDetail)}`);
    }

    return response.blob();
  }

  /**
   * Export backtest results in the requested format.
   * @param {number} backtestId - Backtest ID
   * @param {string} format - Export format
   * @param {boolean} includeTransactions - Include transaction details
   * @param {boolean} includePositions - Include position snapshots
   * @returns {Promise<Blob>} - Exported data
   */
  async exportBacktestResults(
    backtestId,
    format = 'json',
    includeTransactions = false,
    includePositions = false,
  ) {
    const url = new URL(`${this.baseUrl}/export/backtest/${backtestId}`);
    url.searchParams.append('format', format);
    url.searchParams.append('include_transactions', includeTransactions);
    url.searchParams.append('include_positions', includePositions);

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    if (!response.ok) {
      let errorDetail = {};
      try {
        errorDetail = await response.json();
      } catch (e) {
        errorDetail = { detail: await response.text() };
      }
      throw new Error(`API error: ${response.status} - ${JSON.stringify(errorDetail)}`);
    }

    return response.blob();
  }

  /**
   * Export strategy details in the requested format.
   * @param {number} strategyId - Strategy ID
   * @param {string} format - Export format
   * @param {boolean} includeCode - Include strategy code
   * @param {boolean} includeBacktests - Include backtest results
   * @returns {Promise<Blob>} - Exported data
   */
  async exportStrategy(
    strategyId,
    format = 'json',
    includeCode = false,
    includeBacktests = false,
  ) {
    const url = new URL(`${this.baseUrl}/export/strategy/${strategyId}`);
    url.searchParams.append('format', format);
    url.searchParams.append('include_code', includeCode);
    url.searchParams.append('include_backtests', includeBacktests);

    const response = await fetch(url, {
      method: 'GET',
      headers: this._getHeaders(),
    });

    if (!response.ok) {
      let errorDetail = {};
      try {
        errorDetail = await response.json();
      } catch (e) {
        errorDetail = { detail: await response.text() };
      }
      throw new Error(`API error: ${response.status} - ${JSON.stringify(errorDetail)}`);
    }

    return response.blob();
  }

  /**
   * Connect to WebSocket for backtest monitoring.
   * @param {string} clientId - Client ID
   * @param {Function} onMessage - Message handler
   * @param {Function} onError - Error handler
   * @param {Function} onClose - Close handler
   * @returns {WebSocket} - WebSocket connection
   */
  connectToBacktestWebSocket(clientId, onMessage, onError, onClose) {
    const url = new URL(`${this.baseUrl.replace('http', 'ws')}/ws/backtest/${clientId}`);
    
    if (this.token) {
      url.searchParams.append('token', this.token);
    }

    const ws = new WebSocket(url);

    ws.onmessage = (event) => {
      try {
        const data = JSON.parse(event.data);
        onMessage(data);
      } catch (e) {
        console.error('Error parsing WebSocket message:', e);
      }
    };

    ws.onerror = (event) => {
      if (onError) {
        onError(event);
      }
    };

    ws.onclose = (event) => {
      if (onClose) {
        onClose(event);
      }
    };

    return ws;
  }

  /**
   * Subscribe to backtest updates.
   * @param {WebSocket} ws - WebSocket connection
   * @param {number} backtestId - Backtest ID
   */
  subscribeToBacktest(ws, backtestId) {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify({
        type: 'subscribe',
        backtest_id: backtestId,
      }));
    }
  }

  /**
   * Unsubscribe from backtest updates.
   * @param {WebSocket} ws - WebSocket connection
   * @param {number} backtestId - Backtest ID
   */
  unsubscribeFromBacktest(ws, backtestId) {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify({
        type: 'unsubscribe',
        backtest_id: backtestId,
      }));
    }
  }

  /**
   * Get current subscriptions.
   * @param {WebSocket} ws - WebSocket connection
   */
  getSubscriptions(ws) {
    if (ws.readyState === WebSocket.OPEN) {
      ws.send(JSON.stringify({
        type: 'get_subscriptions',
      }));
    }
  }
}

// Export for Node.js or browser
if (typeof module !== 'undefined' && module.exports) {
  module.exports = USDCArbitrageClient;
} else {
  window.USDCArbitrageClient = USDCArbitrageClient;
}

File: src/api/client/python/usdc_arbitrage_client.py
"""Python client SDK for USDC Arbitrage API."""

import json
import logging
from datetime import datetime
from typing import Any, Dict, List, Optional, Union

import requests
from pydantic import BaseModel, Field
from requests.auth import AuthBase


class TokenAuth(AuthBase):
    """Token authentication for requests."""

    def __init__(self, token: str):
        self.token = token

    def __call__(self, r):
        r.headers["Authorization"] = f"Bearer {self.token}"
        return r


class BacktestRequest(BaseModel):
    """Backtest request model."""

    strategy_id: int = Field(..., description="Strategy ID to backtest")
    start_date: datetime = Field(..., description="Start date for backtest")
    end_date: datetime = Field(..., description="End date for backtest")
    exchanges: List[str] = Field(
        default=["coinbase", "kraken", "binance"],
        description="Exchanges to include in backtest",
    )
    symbols: List[str] = Field(
        default=["USDC/USD"],
        description="Symbols to include in backtest",
    )
    timeframe: str = Field(
        default="1h",
        description="Timeframe for backtest (e.g., 1m, 5m, 1h, 1d)",
    )
    initial_balance: float = Field(
        default=10000.0,
        description="Initial portfolio balance",
        gt=0,
    )
    position_sizing: str = Field(
        default="percent",
        description="Position sizing strategy (fixed, percent, kelly, volatility, risk_parity)",
    )
    position_size: float = Field(
        default=0.02,
        description="Position size (interpretation depends on position_sizing)",
        gt=0,
    )
    rebalance_frequency: str = Field(
        default="monthly",
        description="Portfolio rebalancing frequency (daily, weekly, monthly, quarterly, threshold)",
    )
    rebalance_threshold: float = Field(
        default=0.05,
        description="Threshold for threshold-based rebalancing",
        gt=0,
    )
    include_fees: bool = Field(
        default=True,
        description="Include exchange fees in backtest",
    )
    include_slippage: bool = Field(
        default=True,
        description="Include slippage in backtest",
    )
    strategy_params: Dict[str, Any] = Field(
        default={},
        description="Additional strategy parameters",
    )


class StrategyCreateRequest(BaseModel):
    """Strategy creation request model."""

    name: str = Field(..., description="Strategy name")
    description: str = Field(..., description="Strategy description")
    code: str = Field(..., description="Strategy code")
    parameters: Dict[str, Any] = Field(default={}, description="Strategy parameters")
    tags: Optional[List[str]] = Field(default=None, description="Strategy tags")
    strategy_type: str = Field(default="custom", description="Strategy type")
    is_active: bool = Field(default=True, description="Whether strategy is active")


class StrategyUpdateRequest(BaseModel):
    """Strategy update request model."""

    name: Optional[str] = Field(default=None, description="Strategy name")
    description: Optional[str] = Field(default=None, description="Strategy description")
    code: Optional[str] = Field(default=None, description="Strategy code")
    parameters: Optional[Dict[str, Any]] = Field(
        default=None, description="Strategy parameters"
    )
    tags: Optional[List[str]] = Field(default=None, description="Strategy tags")
    is_active: Optional[bool] = Field(
        default=None, description="Whether strategy is active"
    )
    commit_message: str = Field(
        default="", description="Commit message for version control"
    )


class USDCArbitrageClient:
    """Client for USDC Arbitrage API."""

    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        api_version: str = "1.0",
        token: Optional[str] = None,
    ):
        """Initialize client."""
        self.base_url = base_url
        self.api_version = api_version
        self.token = token
        self.logger = logging.getLogger(__name__)

    def _get_headers(self) -> Dict[str, str]:
        """Get request headers."""
        headers = {
            "Content-Type": "application/json",
            "Accept": "application/json",
            "Accept-Version": self.api_version,
        }

        if self.token:
            headers["Authorization"] = f"Bearer {self.token}"

        return headers

    def _handle_response(self, response: requests.Response) -> Dict[str, Any]:
        """Handle API response."""
        try:
            response.raise_for_status()
            return response.json()
        except requests.exceptions.HTTPError as e:
            self.logger.error(f"HTTP error: {e}")
            error_detail = {}
            try:
                error_detail = response.json()
            except:
                error_detail = {"detail": response.text}

            raise Exception(f"API error: {response.status_code} - {error_detail}")
        except requests.exceptions.RequestException as e:
            self.logger.error(f"Request error: {e}")
            raise Exception(f"Request error: {e}")
        except json.JSONDecodeError:
            self.logger.error("Failed to decode JSON response")
            raise Exception("Failed to decode JSON response")

    def login(self, username: str, password: str) -> Dict[str, Any]:
        """Login to API and get access token."""
        url = f"{self.base_url}/auth/token"
        data = {
            "username": username,
            "password": password,
        }

        response = requests.post(url, json=data, headers=self._get_headers())
        result = self._handle_response(response)

        if "access_token" in result:
            self.token = result["access_token"]

        return result

    def refresh_token(self, refresh_token: str) -> Dict[str, Any]:
        """Refresh access token."""
        url = f"{self.base_url}/auth/refresh"
        data = {
            "refresh_token": refresh_token,
        }

        response = requests.post(url, json=data, headers=self._get_headers())
        result = self._handle_response(response)

        if "access_token" in result:
            self.token = result["access_token"]

        return result

    # Strategy endpoints
    def list_strategies(
        self,
        active_only: bool = False,
        tag: Optional[str] = None,
        search: Optional[str] = None,
        skip: int = 0,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """List strategies."""
        url = f"{self.base_url}/strategies/"
        params = {
            "active_only": active_only,
            "skip": skip,
            "limit": limit,
        }

        if tag:
            params["tag"] = tag

        if search:
            params["search"] = search

        response = requests.get(url, params=params, headers=self._get_headers())
        return self._handle_response(response)

    def get_strategy(
        self, strategy_id: int, include_code: bool = False
    ) -> Dict[str, Any]:
        """Get strategy by ID."""
        url = f"{self.base_url}/strategies/{strategy_id}"
        params = {
            "include_code": include_code,
        }

        response = requests.get(url, params=params, headers=self._get_headers())
        return self._handle_response(response)

    def create_strategy(
        self, request: Union[StrategyCreateRequest, Dict[str, Any]], author: str = "api"
    ) -> Dict[str, Any]:
        """Create a new strategy."""
        url = f"{self.base_url}/strategies/"
        params = {
            "author": author,
        }

        if isinstance(request, StrategyCreateRequest):
            data = request.model_dump()
        else:
            data = request

        response = requests.post(
            url, params=params, json=data, headers=self._get_headers()
        )
        return self._handle_response(response)

    def update_strategy(
        self,
        strategy_id: int,
        request: Union[StrategyUpdateRequest, Dict[str, Any]],
        author: str = "api",
    ) -> Dict[str, Any]:
        """Update an existing strategy."""
        url = f"{self.base_url}/strategies/{strategy_id}"
        params = {
            "author": author,
        }

        if isinstance(request, StrategyUpdateRequest):
            data = request.model_dump(exclude_unset=True)
        else:
            data = request

        response = requests.put(
            url, params=params, json=data, headers=self._get_headers()
        )
        return self._handle_response(response)

    def delete_strategy(self, strategy_id: int) -> Dict[str, Any]:
        """Delete a strategy."""
        url = f"{self.base_url}/strategies/{strategy_id}"

        response = requests.delete(url, headers=self._get_headers())
        return self._handle_response(response)

    # Backtest endpoints
    def run_backtest(
        self, request: Union[BacktestRequest, Dict[str, Any]]
    ) -> Dict[str, Any]:
        """Run a backtest."""
        url = f"{self.base_url}/backtest/"

        if isinstance(request, BacktestRequest):
            data = request.model_dump()
        else:
            data = request

        response = requests.post(url, json=data, headers=self._get_headers())
        return self._handle_response(response)

    def get_backtest(self, backtest_id: int) -> Dict[str, Any]:
        """Get backtest by ID."""
        url = f"{self.base_url}/backtest/{backtest_id}"

        response = requests.get(url, headers=self._get_headers())
        return self._handle_response(response)

    def list_backtests(
        self,
        strategy_id: Optional[int] = None,
        status: Optional[str] = None,
        limit: int = 10,
        offset: int = 0,
    ) -> List[Dict[str, Any]]:
        """List backtests."""
        url = f"{self.base_url}/backtest/"
        params = {
            "limit": limit,
            "offset": offset,
        }

        if strategy_id is not None:
            params["strategy_id"] = strategy_id

        if status is not None:
            params["status"] = status

        response = requests.get(url, params=params, headers=self._get_headers())
        return self._handle_response(response)

    # Results endpoints
    def list_results(
        self,
        strategy_id: Optional[int] = None,
        status: Optional[str] = None,
        skip: int = 0,
        limit: int = 100,
        sort_by: str = "created_at",
        sort_order: str = "desc",
    ) -> List[Dict[str, Any]]:
        """List backtest results."""
        url = f"{self.base_url}/results/"
        params = {
            "skip": skip,
            "limit": limit,
            "sort_by": sort_by,
            "sort_order": sort_order,
        }

        if strategy_id is not None:
            params["strategy_id"] = strategy_id

        if status is not None:
            params["status"] = status

        response = requests.get(url, params=params, headers=self._get_headers())
        return self._handle_response(response)

    def get_result(self, result_id: int) -> Dict[str, Any]:
        """Get backtest result by ID."""
        url = f"{self.base_url}/results/{result_id}"

        response = requests.get(url, headers=self._get_headers())
        return self._handle_response(response)

    def get_result_transactions(
        self, result_id: int, skip: int = 0, limit: int = 100
    ) -> List[Dict[str, Any]]:
        """Get transactions for a backtest result."""
        url = f"{self.base_url}/results/{result_id}/transactions"
        params = {
            "skip": skip,
            "limit": limit,
        }

        response = requests.get(url, params=params, headers=self._get_headers())
        return self._handle_response(response)

    def get_result_positions(
        self, result_id: int, skip: int = 0, limit: int = 100
    ) -> List[Dict[str, Any]]:
        """Get position snapshots for a backtest result."""
        url = f"{self.base_url}/results/{result_id}/positions"
        params = {
            "skip": skip,
            "limit": limit,
        }

        response = requests.get(url, params=params, headers=self._get_headers())
        return self._handle_response(response)

    def get_result_equity_curve(self, result_id: int) -> List[Dict[str, Any]]:
        """Get equity curve for a backtest result."""
        url = f"{self.base_url}/results/{result_id}/equity_curve"

        response = requests.get(url, headers=self._get_headers())
        return self._handle_response(response)

    # Data export endpoints
    def export_data(
        self,
        data_type: str,
        format: str = "json",
        filters: Optional[Dict[str, Any]] = None,
        start_date: Optional[datetime] = None,
        end_date: Optional[datetime] = None,
        exchanges: Optional[List[str]] = None,
        symbols: Optional[List[str]] = None,
        strategy_id: Optional[int] = None,
        backtest_id: Optional[int] = None,
        limit: Optional[int] = 10000,
    ) -> bytes:
        """Export data in the requested format."""
        url = f"{self.base_url}/export/data"
        data = {
            "data_type": data_type,
            "format": format,
        }

        if filters:
            data["filters"] = filters

        if start_date:
            data["start_date"] = start_date.isoformat()

        if end_date:
            data["end_date"] = end_date.isoformat()

        if exchanges:
            data["exchanges"] = exchanges

        if symbols:
            data["symbols"] = symbols

        if strategy_id:
            data["strategy_id"] = strategy_id

        if backtest_id:
            data["backtest_id"] = backtest_id

        if limit:
            data["limit"] = limit

        response = requests.post(url, json=data, headers=self._get_headers())

        try:
            response.raise_for_status()
            return response.content
        except requests.exceptions.HTTPError as e:
            self.logger.error(f"HTTP error: {e}")
            error_detail = {}
            try:
                error_detail = response.json()
            except:
                error_detail = {"detail": response.text}

            raise Exception(f"API error: {response.status_code} - {error_detail}")
        except requests.exceptions.RequestException as e:
            self.logger.error(f"Request error: {e}")
            raise Exception(f"Request error: {e}")

    def export_backtest_results(
        self,
        backtest_id: int,
        format: str = "json",
        include_transactions: bool = False,
        include_positions: bool = False,
    ) -> bytes:
        """Export backtest results in the requested format."""
        url = f"{self.base_url}/export/backtest/{backtest_id}"
        params = {
            "format": format,
            "include_transactions": include_transactions,
            "include_positions": include_positions,
        }

        response = requests.get(url, params=params, headers=self._get_headers())

        try:
            response.raise_for_status()
            return response.content
        except requests.exceptions.HTTPError as e:
            self.logger.error(f"HTTP error: {e}")
            error_detail = {}
            try:
                error_detail = response.json()
            except:
                error_detail = {"detail": response.text}

            raise Exception(f"API error: {response.status_code} - {error_detail}")
        except requests.exceptions.RequestException as e:
            self.logger.error(f"Request error: {e}")
            raise Exception(f"Request error: {e}")

    def export_strategy(
        self,
        strategy_id: int,
        format: str = "json",
        include_code: bool = False,
        include_backtests: bool = False,
    ) -> bytes:
        """Export strategy details in the requested format."""
        url = f"{self.base_url}/export/strategy/{strategy_id}"
        params = {
            "format": format,
            "include_code": include_code,
            "include_backtests": include_backtests,
        }

        response = requests.get(url, params=params, headers=self._get_headers())

        try:
            response.raise_for_status()
            return response.content
        except requests.exceptions.HTTPError as e:
            self.logger.error(f"HTTP error: {e}")
            error_detail = {}
            try:
                error_detail = response.json()
            except:
                error_detail = {"detail": response.text}

            raise Exception(f"API error: {response.status_code} - {error_detail}")
        except requests.exceptions.RequestException as e:
            self.logger.error(f"Request error: {e}")
            raise Exception(f"Request error: {e}")
